src/v/storage/types.h

```diff
@@ -93,7 +103,17 @@ class snapshotable_stm {
  */
 class stm_manager {
 public:
-    void add_stm(ss::shared_ptr<snapshotable_stm> stm) { _stms.push_back(stm); }
+    template<
+      typename T,
+      typename = std::enable_if_t<std::is_base_of_v<snapshotable_stm, T>>>
+    void add_stm(ss::shared_ptr<T> stm) {
```

Comment:
It might be a clearer api if you added an `add_compactable_stm` interface since we already have a run-time requirement that two compatible stms cannot be added

Reply:
Agreed.

Reply:
I ended up not doing this (separately-named method) because of the awkwardness of the class hierarchy; a `compactable_stm` is not a `snapshotable_stm`. Ended up using a dynamic cast. LMK what you think.